### PR TITLE
Add a button to send edit link by email in anonymous mode

### PR DESCRIPTION
**Pull request by yohanboniface**
This is a work I started last year, and not yet finished.

The goal is to add a button "Send me the link" when creating a map in anonymous mode.

### CODE REVIEW

File: umap/static/umap/js/umap.js [hunk 1]
```diff
@@ -1336,11 +1336,23 @@ L.U.Map.include({
...
         if (!this.options.umap_id) {
-          duration = 100000 // we want a longer message at map creation (TODO UGLY)
+          alert.duration = 100000 // we want a longer message at map creation (TODO UGLY)
           this.options.umap_id = data.id
           this.permissions.setOptions(data.permissions)
+          if (data.permissions && data.permissions.anonymous_edit_url) {
+            alert.actions = [
+              {
+                label: L._('Send me edit link by email'),
+                callback: function () {
```
Comment (davidbgk) [inline umap/static/umap/js/umap.js:1349]:
```suggestion
                callback: this.sendEditLink
```

Might work?

File: umap/static/umap/js/umap.js [hunk 2]
```diff
@@ -1350,18 +1362,23 @@ L.U.Map.include({
...
+          this.ui.alert(alert)
         })
         this.ui.closePanel()
         this.permissions.save()
       },
     })
   },
 
+  sendEditLink: function () {
+    var url = L.Util.template(this.options.urls.map_send_edit_link, {
```
Comment (davidbgk) [inline umap/static/umap/js/umap.js:1377]: `const` 😛

File: umap/views.py [hunk 3]
```diff
@@ -622,6 +622,28 @@ def post(self, *args, **kwargs):
...
+    model = Map
+    pk_url_kwarg = 'map_id'
+
+    def form_valid(self, form):
+        if (self.object.owner
+           or not self.object.is_anonymous_owner(self.request)
+           or not self.object.can_edit(self.request.user, self.request)):
+            return HttpResponseForbidden()
+        email = form.cleaned_data["email"]
+        from django.core.mail import send_mail
```
Comment (davidbgk) [inline umap/views.py:635]: Does `send_mail` import requires external dependencies/settings? If so, maybe add a comment?

**Comment by yohanboniface** [comment]
We should have a working version, now let's make it clean :p

image

File: umap/views.py (hunk 3 above)
Comment (yohanboniface) [reply umap/views.py]: Yes, settings. I'll add an entry in the documentation.

File: umap/static/umap/js/umap.js (hunk 1 above)
Comment (yohanboniface) [reply umap/static/umap/js/umap.js]: Yes! Done :)

**Comment by davidbgk** [comment]
Is that possible to make the link clickable? Last time I tried I failed (because it was not on the top layer?)

**Comment by yohanboniface** [comment]
If you click it, you'll lose it :p

It will open a page that will add a cookie (which is already present in this case, given the map has just been created), then redirect to the map with the normal URL.

**Comment by davidbgk** [comment]
Maybe a `Copy link` button close to it then?

**Comment by yohanboniface** [comment]
image

Simpler than expected!

But it will work only in https, right ?

**Comment by davidbgk** [comment]
> But it will work only in https, right ?

I don't think so, did you use the [Navigator.clipboard](https://developer.mozilla.org/en-US/docs/Web/API/Navigator/clipboard) API? (e.g. https://code.larlet.fr/javascript/#copier-dans-le-presse-papier)

**Comment by yohanboniface** [comment]
> I don't think so, did you use the [Navigator.clipboard](https://developer.mozilla.org/en-US/docs/Web/API/Navigator/clipboard) API

Yep, see last commit :)

File: umap/static/umap/js/umap.ui.js [hunk 4]
```diff
@@ -108,26 +107,33 @@ L.U.UI = L.Evented.extend({
...
         if (action.callback)
-          L.DomEvent.on(
-            el,
-            'click',
-            action.callback,
-            action.callbackContext || this.map
-          )
+          L.DomEvent.on(el, 'click', () => {
+            action.callback.bind(action.callbackContext || this.map)()
+            close()
```
Comment (yohanboniface) [inline umap/static/umap/js/umap.ui.js:127]: This `close` call does not work :/

File: umap/settings/base.py [hunk 5]
```diff
@@ -114,6 +114,9 @@
 )
 DEFAULT_AUTO_FIELD = 'django.db.models.AutoField'
 
+EMAIL_BACKEND = "django.core.mail.backends.console.EmailBackend"
+FROM_EMAIL = None
```
Comment (davidbgk) [inline umap/settings/base.py:118]: I still wonder if we should add `FROM_EMAIL = "test@example.com"` to `local.py.sample` to see the feature in action when developing or if that's really an on-demand one 🤔

File: umap/static/umap/js/umap.js [hunk 6]
```diff
@@ -1336,11 +1336,40 @@ L.U.Map.include({
...
+              {
+                label: L._('Send me the link'),
+                input: L._('Email'),
+                callback: this.sendEditLink,
+                callbackContext: this,
+              },
+              {
+                label: L._('Copy link'),
+                callback: () => {
+                  navigator.clipboard.writeText(data.permissions.anonymous_edit_url)
```
[the remark this answers is not in the document]
Comment (yohanboniface) [reply umap/static/umap/js/umap.js:1366]: Or is that the sign that we should force https ?

File: umap/views.py [hunk 7]
```diff
@@ -622,6 +611,36 @@ def post(self, *args, **kwargs):
...
+            return HttpResponseForbidden()
+        form = self.get_form()
+        if form.is_valid():
+            email = form.cleaned_data["email"]
+        else:
+            return HttpResponseBadRequest("Invalid")
+        link = self.object.get_anonymous_edit_url()
+
+        send_mail(
+            _("Your secret edit link"),
```
Comment (davidbgk) [inline umap/views.py:633]:
```suggestion
            _("The uMap edit link for your map: %(map_name)s" % {"map_name": self.object.name}),
```

This original subject looks spam-y to me :)

**Review by davidbgk — APPROVED** [review]

**Pull request merged**